cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,314 @@
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	if o.ciOperatorconfigPath == "" {
+		return fmt.Errorf("--ci-operator-config-path must be set")
+	}
+	if o.releaseControllerMirrorConfigDir == "" {
+		return fmt.Errorf("--release-controller-mirror-config-dir must be set")
+	}
+	for _, s := range o.ignoredImageStreamTagsRaw.Strings() {
+		re, err := regexp.Compile(s)
+		if err != nil {
+			return fmt.Errorf("failed to compile regex from %q: %w", s, err)
+		}
+		o.ignoredImageStreamTags = append(o.ignoredImageStreamTags, re)
+	}
+	return nil
+}
+
+func addSchemes() error {
+	if err := imagev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add imagev1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func validatePromotedTags(tags []api.ImageStreamTagReference) error {
+	tagCount := map[api.ImageStreamTagReference]int{}
+	for _, tag := range tags {
+		if _, ok := tagCount[tag]; ok {
+			tagCount[tag] = tagCount[tag] + 1
+		} else {
+			tagCount[tag] = 1
+		}
+	}
+	var dups []string
+	for tag, c := range tagCount {
+		if c > 1 {
+			dups = append(dups, tag.ISTagName())
+		}
+	}
+	if len(dups) > 0 {
+		return fmt.Errorf("found tags promoted more than once: %s", dups)
+	}
+	return nil
+}
+
+func tagsToDelete(ctx context.Context, client ctrlruntimeclient.Client, promotedTags []api.ImageStreamTagReference, toIgnore []*regexp.Regexp, imageStreamRefs []ImageStreamRef) ([]api.ImageStreamTagReference, error) {
+	imageStreamsWithPromotedTags := map[ctrlruntimeclient.ObjectKey]interface{}{}
+	for _, promotedTag := range promotedTags {
+		imageStreamsWithPromotedTags[ctrlruntimeclient.ObjectKey{Namespace: promotedTag.Namespace, Name: promotedTag.Name}] = nil
+	}
+
+	tagsToCheck := map[api.ImageStreamTagReference]interface{}{}
+	var errs []error
+	for objectKey := range imageStreamsWithPromotedTags {
+		imageStream := &imagev1.ImageStream{}
+		if err := client.Get(ctx, objectKey, imageStream); err != nil {
+			if !kerrors.IsNotFound(err) {
+				errs = append(errs, fmt.Errorf("could not get image stream %s in namespace %s", objectKey.Name, objectKey.Namespace))
+			}
+			continue
+		}
+		for _, tag := range imageStream.Status.Tags {
+			tagsToCheck[api.ImageStreamTagReference{Namespace: imageStream.Namespace, Name: imageStream.Name, Tag: tag.Tag}] = nil
+		}
+	}
+
+	for _, promotedTag := range promotedTags {
+		delete(tagsToCheck, promotedTag)
+	}
+	for tag := range tagsToCheck {
+		for _, re := range toIgnore {
+			if re.MatchString(tag.ISTagName()) {
+				delete(tagsToCheck, tag)
+			}
+		}
+	}
+	mirroredTags, err := mirroredTagsByReleaseController(ctx, client, imageStreamRefs)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to get the mirrored tags by release controller: %w", err))
+	}
+	for _, tag := range mirroredTags {
+		delete(tagsToCheck, tag)
+	}
+	if len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	var tagsToDelete []api.ImageStreamTagReference
```

Comment:
What is the point of copying the data from `tagsToCheck` into this when they are equal?

Reply:
Removed.
Returned the map instead.